pkg/metrics/controller.go

```diff
@@ -250,7 +259,64 @@ func (c *Controller) recordFleetDeletion(obj interface{}) {
 		return
 	}
 
-	c.recordFleetReplicas(f.Name, f.Namespace, 0, 0, 0, 0, 0)
+	if runtime.FeatureEnabled(runtime.FeatureResetMetricsOnDelete) {
+		if err := c.resyncFleets(); err != nil {
+			// If for some reason resync fails, the entire metric state for fleets
+			// will be reset whenever the next Fleet gets deleted, in which case
+			// we end up back in a healthy state - so we aren't going to actively retry.
+			c.logger.WithError(err).Warn("Could not resync Fleet Metrics")
+		}
+	} else {
+		c.recordFleetReplicas(f.Name, f.Namespace, 0, 0, 0, 0, 0)
+	}
+}
+
+// resyncFleets resets all views associated with a Fleet, and recalculates all totals.
+func (c *Controller) resyncFleets() error {
+	c.lock.Lock()
+	defer c.lock.Unlock()
+	fleets, err := c.fleetLister.List(labels.Everything())
+	if err != nil {
+		return errors.Wrap(err, "could not resync Fleets")
+	}
+
+	fasList, err := c.fasLister.List(labels.Everything())
+	if err != nil {
+		return errors.Wrap(err, "could not resync Fleets")
+	}
+
+	resetViews(fleetReplicaCountName, fleetAutoscalerBufferLimitName, fleetAutoscalterBufferSizeName,
+		fleetAutoscalerCurrentReplicaCountName, fleetAutoscalersDesiredReplicaCountName, fleetAutoscalersAbleToScaleName,
+		fleetAutoscalersLimitedName, gameServersCountName, gameServersTotalName, gameServerStateDurationName)
+	for _, f := range fleets {
+		c.recordFleetChanges(f)
+	}
+	for _, fas := range fasList {
+		c.recordFleetAutoScalerChanges(nil, fas)
+	}
+	c.collectGameServerCounts()
+
+	return nil
+}
+
+// resyncFleetAutoScaler resets all views associated with FleetAutoscalers, and recalculates metric totals.
+func (c *Controller) resyncFleetAutoScaler() error {
```

Comment:
This felt tricky, because passing values on when to lock or not etc seems messy.

But I think I have a good solution - it feels like the biggest potential for error is in which views are associated with a Fleet or a FleetAutoscaler -- so I moved them into variables, and made one reference the other (Fleet is a super set of FleetAutoscaler's list).

That way we have the building blocks, and if new views become available down the line, there is only one place that need to be updated.

PTAL, let me know what you think.

Reply:
The superset thing confused me a bit (see my above comment). 

What's really happening is that you can just reset fleet autoscalers (here), or you can reset fleets and fleet autoscalers (the above function). So only reseting the fleet autoscalers is essentially an optimization for the case where an autoscaler is deleted as you don't need to scan all of the fleets as well. But if a fleet is deleted we want to update the fleet autoscaler metrics in addition to the fleet metrics? Is that necessary?

Reply:
That makes senes. Thanks for explaining.